tokio-postgres/src/connect.rs

```diff
@@ -83,6 +132,433 @@ where
+            let host_to_port_map = HOST_TO_PORT_MAP.lock().unwrap().clone();
+            let mut index = 0;
+            while index < host_list.len() {
+                let host = host_list.get(index);
+                let mut conn_host = host.unwrap().to_owned();
+                //check if we are to use public hosts
+                if USE_PUBLIC_IP.fetch_or(false, Ordering::SeqCst) {
+                    let public_host_map = PUBLIC_HOST_MAP.lock().unwrap().clone();
+                    let public_host = public_host_map.get(&conn_host.clone());
+                    if public_host.is_none() {
+                        info!("Public host not available for private host {:?}, adding this to failed host list and trying another server", conn_host.clone());
+                        add_to_failed_host_list(host.cloned().unwrap());
+                        index = index + 1;
+                        continue;
+                    } else {
+                        conn_host = public_host.unwrap().clone();
+                    }
+                }
+
+                // The value of host is used as the hostname for TLS validation,
+                let hostname = match conn_host.clone() {
+                    Host::Tcp(host) => Some(host),
+                    // postgres doesn't support TLS over unix sockets, so the choice here doesn't matter
+                    #[cfg(unix)]
+                    Host::Unix(_) => None,
+                };
+
+                if let Ok((client, connection)) = connect_host(
+                    conn_host.clone(),
+                    hostname.clone(),
+                    host_to_port_map[&(conn_host.clone())],
+                    &mut NoTls,
+                    config,
+                )
+                .await
+                {
+                    let handle = tokio::spawn(async move {
+                        if let Err(e) = connection.await {
+                            eprintln!("connection error: {}", e);
+                        }
+                    });
+                    info!("Control connection created to {:?}", hostname.clone());
+                    refresh(client, config).await;
+                    handle.abort();
+                    return true;
+                } else {
+                    info!("Failed to establish control connection to {:?}, adding this to failed host list and trying another server", hostname.clone());
+                    add_to_failed_host_list(host.cloned().unwrap());
+                    index = index + 1;
+                }
+            }
+            info!("Failed to establish control connection to available servers");
+            return false;
+        }
+    }
+    return true;
+}
+
+fn add_to_failed_host_list(host: Host) {
+    let mut failedhostlist = FAILED_HOSTS.lock().unwrap();
+    failedhostlist.insert(host.clone(), Instant::now());
+    info!("Added {:?} to failed host list", host.clone());
+}
+
+async fn refresh(client: Client, config: &Config) {
```

Comment:
`refresh()` should be called by only one thread, if there are multiple threads trying to create connections. How is it ensured here?